src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.Multiplexer.cs

```diff
@@ -0,0 +1,490 @@
+
+                            createdRemainderPacket = true;
+
+                            recurse = SetupRemainderPacket(remainderPacket);
+                        }
+                        else
+                        {
+                            newDataLength = TdsEnums.HEADER_LEN + packetDataLength;
+                            int remainderLength = data.Length - (TdsEnums.HEADER_LEN + packetDataLength);
+                            remainderPacket = new Packet
+                            {
+                                Buffer = new byte[dataBuffer.Length],
+                                CurrentLength = remainderLength
+                            };
+                            remainderPacket.SetCreatedBy(5);
+
+                            ReadOnlySpan<byte> remainderSource = data.Slice(TdsEnums.HEADER_LEN + packetDataLength);
+                            Span<byte> remainderTarget = remainderPacket.Buffer.AsSpan(0, remainderLength);
+                            remainderSource.CopyTo(remainderTarget);
+#if DEBUG
+                            // for debugging purposes fill the removed data area with an easily
+                            // recognisable pattern so we can see if it is misused
+                            Span<byte> removed = dataBuffer.AsSpan(TdsEnums.HEADER_LEN + packetDataLength, remainderLength);
+                            removed.Fill(0xFF);
+#endif
+                            createdRemainderPacket = true;
+                            recurse = SetupRemainderPacket(remainderPacket);
+
+                            consumeInputDirectly = true;
+                        }
+                    }
+                }
+                else
+                {
+                    // either:
+                    // 1) we took some data from the input to reconstruct the partial packet
+                    // 2) there was less than a single packet header of data received
+                    // in both cases we can't tell the caller to directly consume the packet
+                    // in the input buffer, we need to construct a new remainder packet with
+                    // the incomplete data and let the caller deal with it
+                    remainderPacket = new Packet
+                    {
+                        Buffer = dataBuffer,
+                        CurrentLength = data.Length
+                    };
+                    remainderPacket.SetCreatedBy(6);
+                    createdRemainderPacket = true;
+                    recurse = SetupRemainderPacket(remainderPacket);
+                }
+            }
+
+            if (consumePartialPacket && consumeInputDirectly)
+            {
+                throw new InvalidOperationException($"MultiplexPackets cannot return both {nameof(consumePartialPacket)} and {nameof(consumeInputDirectly)}");
+            }
+        }
+
+        private static bool SetupRemainderPacket(Packet packet)
```

Comment:
The setter is used in 3 places and each is a different situation. In one case we know the value ahead of time, in the other two we don't know it and we're setting it using the function to fetch it from the header and in theory we can simply call the function for one of those but it'll do extra work that we don't need.

In general the `Packet` class is a dumb container of related data rather than being something that contains interesting logic for behaviours. Accessors has sanity checks but little more. Think of internals more like C than C#. Of course I can move behaviours into it but it increases cognitive load for little payback.